src/Interpreters/Cache/LRUFileCachePriority.cpp

```diff
@@ -166,7 +176,16 @@ bool LRUFileCachePriority::LRUIterator::operator ==(const LRUIterator & other) c
 
 void LRUFileCachePriority::iterate(IterateFunc func, const CachePriorityGuard::Lock & lock)
 {
-    for (auto it = queue.begin(); it != queue.end();)
+    iterateImpl(queue.begin(), func, lock);
+}
+
+LRUFileCachePriority::LRUQueue::iterator
+LRUFileCachePriority::iterateImpl(LRUQueue::iterator start_pos, IterateFunc func, const CachePriorityGuard::Lock & lock)
+{
+    if (start_pos == queue.end())
+        start_pos = queue.begin();
+
+    for (auto it = start_pos; it != queue.end();)
```

Comment:
Maybe we can try doing some cyclic iteration?

```suggestion
    if (queue.empty())
         return queue.end();
    auto it = start_pos;
    do
    {
        if (it == queue.end())
            it = queue.begin();
        ....
    } while (it != start_pos);
```